### PR TITLE
Fix continous patching

### LINKED ISSUE #198: The resources (services and statefulsets) are modified in each iteration of the reconcile loop
**What version of redis operator are you using?**

redis-operator version: 0.9.0

**Does this issue reproduce with the latest release?**

Yes, we try with the master branch at the issue persists

**What operating system and processor architecture are you using (`kubectl version`)?**

Server Version: version.Info{Major:"1", Minor:"20+", GitVersion:"v1.20.11-gke.1300", GitCommit:"9352724e36d9c7c35609d30b2696b0de61d23aa0", GitTreeState:"clean", BuildDate:"2021-10-06T09:26:56Z", GoVersion:"go1.15.15b5", Compiler:"gc", Platform:"linux/amd64"}

**What did you do?**

After install the operator inside a cluster with istio installed and when create many redis resources, we notice that istio increases cpu consumption. We reviewed the reason and saw that the operator is updating the resources of the services in each iteration of the reconcile loop. 

Reviewing the code, we conclude that the banzaicloud / k8s-objectmatcher library always sees modifications in the resources, so they are always updated. We add more tracing, and confirm this point. The patch is never empty:

```
2022-01-06T21:56:56.226+0100	INFO	controller_redis	Syncing Redis StatefulSet with defined properties	{"Request.StateFulSet.Namespace": "test", "Request.StateFulSet.Name": "test-redis", "patch": "{\"apiVersion\":\"apps/v1\",\"kind\":\"StatefulSet\",\"metadata\":{\"creationTimestamp\":null,\"managedFields\":null,\"resourceVersion\":null},\"spec\":{\"volumeClaimTemplates\":[{\"metadata\":{\"name\":\"test-redis\"},\"spec\":{\"accessModes\":[\"ReadWriteOnce\"],\"resources\":{\"requests\":{\"storage\":\"1Gi\"}}},\"status\":{}}]},\"status\":{\"replicas\":0}}"}
2022-01-06T21:56:56.442+0100	INFO	controller_redis	Syncing Redis service with defined properties	{"Request.Service.Namespace": "test", "Request.Service.Name": "test-redis-headless", "patch": "{\"apiVersion\":\"core/v1\",\"kind\":\"Service\",\"metadata\":{\"annotations\":{\"cloud.google.com/neg\":null},\"creationTimestamp\":null,\"managedFields\":null,\"resourceVersion\":null}}"}
2022-01-06T21:56:56.649+0100	INFO	controller_redis	Syncing Redis service with defined properties	{"Request.Service.Namespace": "test", "Request.Service.Name": "test-redis", "patch": "{\"apiVersion\":\"core/v1\",\"kind\":\"Service\",\"metadata\":{\"annotations\":{\"cloud.google.com/neg\":null},\"creationTimestamp\":null,\"managedFields\":null,\"resourceVersion\":null},\"spec\":{\"clusterIP\":null}}"}
2022-01-06T21:57:06.840+0100	INFO	controller_redis	Syncing Redis StatefulSet with defined properties	{"Request.StateFulSet.Namespace": "test", "Request.StateFulSet.Name": "test-redis", "patch": "{\"apiVersion\":\"apps/v1\",\"kind\":\"StatefulSet\",\"metadata\":{\"creationTimestamp\":null,\"managedFields\":null,\"resourceVersion\":null},\"spec\":{\"volumeClaimTemplates\":[{\"metadata\":{\"name\":\"test-redis\"},\"spec\":{\"accessModes\":[\"ReadWriteOnce\"],\"resources\":{\"requests\":{\"storage\":\"1Gi\"}}},\"status\":{}}]},\"status\":{\"replicas\":0}}"}
2022-01-06T21:57:07.036+0100	INFO	controller_redis	Syncing Redis service with defined properties	{"Request.Service.Namespace": "test", "Request.Service.Name": "test-redis-headless", "patch": "{\"apiVersion\":\"core/v1\",\"kind\":\"Service\",\"metadata\":{\"annotations\":{\"cloud.google.com/neg\":null},\"creationTimestamp\":null,\"managedFields\":null,\"resourceVersion\":null}}"}
2022-01-06T21:57:07.228+0100	INFO	controller_redis	Syncing Redis service with defined properties	{"Request.Service.Namespace": "test", "Request.Service.Name": "test-redis", "patch": "{\"apiVersion\":\"core/v1\",\"kind\":\"Service\",\"metadata\":{\"annotations\":{\"cloud.google.com/neg\":null},\"creationTimestamp\":null,\"managedFields\":null,\"resourceVersion\":null},\"spec\":{\"clusterIP\":null}}"}
```

**What did you expect to see?**

We expect that the resources are updating only when needed

**What did you see instead?**

The resources are updating ever and ever, so the istio cluster must recalculate the topology of the mesh again and again.

P.D: add a parameter to increase the reconcile loop duration (not fixed to 10 seconds) is a extra point.

thanks !
j

### PR DESCRIPTION
**Pull request by davidcollom**
This fixes #198 where by they are continuously updated.

```
1.6425249063840268e+09	INFO	controllers.RedisCluster	Reconciling opstree redis Cluster controller	{"Request.Namespace": "default", "Request.Name": "redis-cluster"}
1.642524906391118e+09	INFO	controller_redis	Redis statefulset get action was successful	{"Request.StateFulSet.Namespace": "default", "Request.StateFulSet.Name": "redis-cluster-leader"}
1.642524906399053e+09	INFO	controller_redis	Changes in statefulset Detected, Updating...	{"Request.StateFulSet.Namespace": "default", "Request.StateFulSet.Name": "redis-cluster-leader", "patch": "{\"apiVersion\":\"apps/v1\",\"kind\":\"StatefulSet\"}"}
1.642524906414318e+09	INFO	controller_redis	Redis stateful successfully updated 	{"Request.StateFulSet.Namespace": "default", "Request.StateFulSet.Name": "redis-cluster-leader"}
1.642524906420856e+09	INFO	controller_redis	Redis service get action is successful	{"Request.Service.Namespace": "default", "Request.Service.Name": "redis-cluster-leader-headless"}
1.642524906422257e+09	INFO	controller_redis	Redis service is already in-sync	{"Request.Service.Namespace": "default", "Request.Service.Name": "redis-cluster-leader-headless"}
1.642524906428809e+09	INFO	controller_redis	Redis service get action is successful	{"Request.Service.Namespace": "default", "Request.Service.Name": "redis-cluster-leader"}
1.642524906429905e+09	INFO	controller_redis	Redis service is already in-sync	{"Request.Service.Namespace": "default", "Request.Service.Name": "redis-cluster-leader"}
1.642524906435884e+09	INFO	controller_redis	Redis PodDisruptionBudget get action was successful	{"Request.PodDisruptionBudget.Namespace": "default", "Request.PodDisruptionBudget.Name": "redis-cluster-leader"}
1.6425249064437351e+09	INFO	controller_redis	Redis statefulset get action was successful	{"Request.StateFulSet.Namespace": "default", "Request.StateFulSet.Name": "redis-cluster-follower"}
1.642524906447791e+09	INFO	controller_redis	Changes in statefulset Detected, Updating...	{"Request.StateFulSet.Namespace": "default", "Request.StateFulSet.Name": "redis-cluster-follower", "patch": "{\"apiVersion\":\"apps/v1\",\"kind\":\"StatefulSet\"}"}
1.642524906461236e+09	INFO	controller_redis	Redis stateful successfully updated 	{"Request.StateFulSet.Namespace": "default", "Request.StateFulSet.Name": "redis-cluster-follower"}
1.64252490646837e+09	INFO	controller_redis	Redis service get action is successful	{"Request.Service.Namespace": "default", "Request.Service.Name": "redis-cluster-follower-headless"}
1.642524906469654e+09	INFO	controller_redis	Redis service is already in-sync	{"Request.Service.Namespace": "default", "Request.Service.Name": "redis-cluster-follower-headless"}
1.642524906478118e+09	INFO	controller_redis	Redis service get action is successful	{"Request.Service.Namespace": "default", "Request.Service.Name": "redis-cluster-follower"}
1.6425249064791229e+09	INFO	controller_redis	Redis service is already in-sync	{"Request.Service.Namespace": "default", "Request.Service.Name": "redis-cluster-follower"}
1.642524906485125e+09	INFO	controller_redis	Redis PodDisruptionBudget get action was successful	{"Request.PodDisruptionBudget.Namespace": "default", "Request.PodDisruptionBudget.Name": "redis-cluster-follower"}
1.64252490649404e+09	INFO	controller_redis	Redis statefulset get action was successful	{"Request.StateFulSet.Namespace": "default", "Request.StateFulSet.Name": "redis-cluster-leader"}
1.64252490650302e+09	INFO	controller_redis	Redis statefulset get action was successful	{"Request.StateFulSet.Namespace": "default", "Request.StateFulSet.Name": "redis-cluster-follower"}
1.642524906503072e+09	INFO	controllers.RedisCluster	Will reconcile redis cluster operator in again 10 seconds	{"Request.Namespace": "default", "Request.Name": "redis-cluster"}
1.642524916503865e+09	INFO	controllers.RedisCluster	Reconciling opstree redis Cluster controller	{"Request.Namespace": "default", "Request.Name": "redis-cluster"}
1.642524916520607e+09	INFO	controller_redis	Redis statefulset get action was successful	{"Request.StateFulSet.Namespace": "default", "Request.StateFulSet.Name": "redis-cluster-leader"}
1.642524916526151e+09	INFO	controller_redis	Changes in statefulset Detected, Updating...	{"Request.StateFulSet.Namespace": "default", "Request.StateFulSet.Name": "redis-cluster-leader", "patch": "{\"apiVersion\":\"apps/v1\",\"kind\":\"StatefulSet\"}"}
1.6425249165436642e+09	INFO	controller_redis	Redis stateful successfully updated 	{"Request.StateFulSet.Namespace": "default", "Request.StateFulSet.Name": "redis-cluster-leader"}
1.642524916551121e+09	INFO	controller_redis	Redis service get action is successful	{"Request.Service.Namespace": "default", "Request.Service.Name": "redis-cluster-leader-headless"}
1.6425249165523598e+09	INFO	controller_redis	Redis service is already in-sync	{"Request.Service.Namespace": "default", "Request.Service.Name": "redis-cluster-leader-headless"}
1.6425249165595148e+09	INFO	controller_redis	Redis service get action is successful	{"Request.Service.Namespace": "default", "Request.Service.Name": "redis-cluster-leader"}
1.6425249165611851e+09	INFO	controller_redis	Redis service is already in-sync	{"Request.Service.Namespace": "default", "Request.Service.Name": "redis-cluster-leader"}
1.64252491656762e+09	INFO	controller_redis	Redis PodDisruptionBudget get action was successful	{"Request.PodDisruptionBudget.Namespace": "default", "Request.PodDisruptionBudget.Name": "redis-cluster-leader"}
1.6425249165751832e+09	INFO	controller_redis	Redis statefulset get action was successful	{"Request.StateFulSet.Namespace": "default", "Request.StateFulSet.Name": "redis-cluster-follower"}
1.642524916577554e+09	INFO	controller_redis	Changes in statefulset Detected, Updating...	{"Request.StateFulSet.Namespace": "default", "Request.StateFulSet.Name": "redis-cluster-follower", "patch": "{\"apiVersion\":\"apps/v1\",\"kind\":\"StatefulSet\"}"}
1.642524916593261e+09	INFO	controller_redis	Redis stateful successfully updated 	{"Request.StateFulSet.Namespace": "default", "Request.StateFulSet.Name": "redis-cluster-follower"}
1.642524916599651e+09	INFO	controller_redis	Redis service get action is successful	{"Request.Service.Namespace": "default", "Request.Service.Name": "redis-cluster-follower-headless"}
1.6425249166014888e+09	INFO	controller_redis	Redis service is already in-sync	{"Request.Service.Namespace": "default", "Request.Service.Name": "redis-cluster-follower-headless"}
1.64252491660789e+09	INFO	controller_redis	Redis service get action is successful	{"Request.Service.Namespace": "default", "Request.Service.Name": "redis-cluster-follower"}
1.642524916608844e+09	INFO	controller_redis	Redis service is already in-sync	{"Request.Service.Namespace": "default", "Request.Service.Name": "redis-cluster-follower"}
1.642524916615288e+09	INFO	controller_redis	Redis PodDisruptionBudget get action was successful	{"Request.PodDisruptionBudget.Namespace": "default", "Request.PodDisruptionBudget.Name": "redis-cluster-follower"}
1.642524916622059e+09	INFO	controller_redis	Redis statefulset get action was successful	{"Request.StateFulSet.Namespace": "default", "Request.StateFulSet.Name": "redis-cluster-leader"}
1.642524916629102e+09	INFO	controller_redis	Redis statefulset get action was successful	{"Request.StateFulSet.Namespace": "default", "Request.StateFulSet.Name": "redis-cluster-follower"}
1.6425249166291358e+09	INFO	controllers.RedisCluster	Will reconcile redis cluster operator in again 10 seconds	{"Request.Namespace": "default", "Request.Name": "redis-cluster"}
```

This is by no means perfect nor ideal...

### CODE REVIEW

File: controllers/rediscluster_controller.go [hunk 1]
```diff
@@ -61,7 +61,7 @@ func (r *RedisClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request
 	if err != nil {
 		return ctrl.Result{}, err
 	}
-	if instance.Spec.RedisLeader.Replicas != nil && *instance.Spec.RedisLeader.Replicas != 0 {
+	if (instance.Spec.Size != nil && *instance.Spec.Size != 0) || (instance.Spec.RedisLeader.Replicas != nil && *instance.Spec.RedisLeader.Replicas != 0) {
```
Comment (davidcollom) [inline controllers/rediscluster_controller.go:64]: If we only used the `instance.Spec.Size` field, we never deployed a service, even though we were deploying equal Follower/Leaders.

File: k8sutils/poddisruption.go [hunk 2]
```diff
@@ -5,7 +5,7 @@ import (
 
 	"github.com/banzaicloud/k8s-objectmatcher/patch"
 	"github.com/go-logr/logr"
-	policyv1beta1 "k8s.io/api/policy/v1beta1"
+	policyv1 "k8s.io/api/policy/v1"
```
Comment (davidcollom) [inline k8sutils/poddisruption.go:8]: Switching policy to V1 from v1beta1 as its due to be removed in `v1.25`

File: k8sutils/services.go [hunk 3]
```diff
@@ -172,16 +176,27 @@ func CreateOrUpdateService(namespace string, serviceMeta metav1.ObjectMeta, owne
...
 	logger := serviceLogger(namespace, storedService.Name)
-	patchResult, err := patch.DefaultPatchMaker.Calculate(storedService, newService, patch.IgnoreStatusFields())
+	// We want to try and keep this atomic as possible.
+	newService.ResourceVersion = storedService.ResourceVersion
+	newService.CreationTimestamp = storedService.CreationTimestamp
+	newService.ManagedFields = storedService.ManagedFields
+
+	if newService.Spec.Type == generateServiceType("ClusterIP") {
+		newService.Spec.ClusterIP = storedService.Spec.ClusterIP
+	}
```
Comment (davidcollom) [inline k8sutils/services.go:186]: Only Set the `spec.ClusterIP` field, if it is of type `ClusterIP`

File: k8sutils/services.go [hunk 4]
```diff
@@ -172,16 +176,27 @@ func CreateOrUpdateService(namespace string, serviceMeta metav1.ObjectMeta, owne
...
+
+	patchResult, err := patch.DefaultPatchMaker.Calculate(storedService, newService,
+		patch.IgnoreStatusFields(),
+		patch.IgnoreField("kind"),
+		patch.IgnoreField("apiVersion"),
```
Comment (davidcollom) [inline k8sutils/services.go:191]: Without these, I was always getting a change, when they were identically set. This removes it from being considered within the patch.

File: k8sutils/statefulset.go [hunk 5]
```diff
@@ -171,12 +176,17 @@ func createPVCTemplate(stsMeta metav1.ObjectMeta, storageSpec corev1.PersistentV
 	} else {
 		pvcTemplate.Spec.AccessModes = storageSpec.Spec.AccessModes
 	}
+	pvcVolumeMode := corev1.PersistentVolumeFilesystem
+	if storageSpec.Spec.VolumeMode != nil {
+		pvcVolumeMode = *storageSpec.Spec.VolumeMode
+	}
```
Comment (davidcollom) [inline k8sutils/statefulset.go:182]: If no VolumeMode is set, assume `FileSystem`... again, not ideal... but if one is not set, we get a constant update.

**Comment by jalonsoa** [comment]
Hi David !,
   thanks a lot for patch ... I'm trying this patch, but I think that is not complete. The statefulset is ok, but the services continue updating all times:

```
2022-01-21T20:54:25.259+0100	INFO	controller_redis	Changes in service Detected, Updating...	{"Request.Service.Namespace": "test", "Request.Service.Name": "test-redis", "patch": "{\"metadata\":{\"annotations\":{\"cloud.google.com/neg\":null,\"meta.helm.sh/release-name\":null,\"redis.opstreelabs.instance\":null}}}"}
```

**Comment by davidcollom** [comment]
@jalonsoa 🤔  I think you might be right... only question would be, are you patching some annotations to the service after creation?
I recognise the `cloud.google.com/neg` annotations, and looking at the patch, its attempting to remove it... if so, I wonder if its worth adding an annotations/labels section for the services within the `Redis`/`RedisCluster` CustomResource...

Also curious on the `meta.helm.sh/release-name` - are you deploying the `Redis`/`RedisCluster` CR/Operator via helm? 🤔 

Happy to implement the changes though.. as we do similar already with the statefulset.

**Comment by jalonsoa** [comment]
> @jalonsoa thinking I think you might be right... only question would be, are you patching some annotations to the service after creation? I recognise the `cloud.google.com/neg` annotations, and looking at the patch, its attempting to remove it... if so, I wonder if its worth adding an annotations/labels section for the services within the `Redis`/`RedisCluster` CustomResource...
> 
> Also curious on the `meta.helm.sh/release-name` - are you deploying the `Redis`/`RedisCluster` CR/Operator via helm? thinking

No, we don't patch the services / statefulset after creation. And yes, we are using jenkins-x v3 to manage our cluster, so we are using helm v3 to deploy the redis-operator

> 
> Happy to implement the changes though.. as we do similar already with the statefulset.

Perfect !! .. I try the redis-operator with this patch and all is working fine for me (we adapt the patch to the v0.9.0 version)

j

**Comment by davidcollom** [comment]
@iamabhishek-dubey Conflicts resolved and ready fro review 🤞

**Review by iamabhishek-dubey — APPROVED** [review]
LGTM, Thanks for this @davidcollom

**Comment by jalonsoa** [comment]
@davidcollom sorry .. but reviewing the patch .. I can't see the change to move the for loop ... So, the patch is not complete. I try the patch (adapted to the v0.9.0) with this change ... At this moment, this patch don't fix the issue for services resources

**Pull request merged**